File: sdks/ts/memory/src/ingest/delta.test.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Unit tests for chunk-level change detection (delta computation and
 * manifest persistence). Mirrors the test coverage in
 * go/knowledge/delta_test.go.
 */

import { afterEach, describe, expect, it } from 'vitest'
import { createMemStore } from '../store/memstore.js'
import type { Store } from '../store/index.js'
import type { Chunk } from './chunker.js'
import {
  buildChunkManifest,
  computeChunkDeltas,
  hashChunk,
  readChunkManifest,
  writeChunkManifest,
} from './delta.js'

const makeChunk = (content: string, ordinal: number): Chunk => ({
  content,
  ordinal,
  headingPath: [],
  startLine: 0,
  endLine: 0,
  tokens: Math.ceil(content.length / 4),
})

describe('hashChunk', () => {
  it('produces deterministic output', () => {
    const h1 = hashChunk('hello world')
    const h2 = hashChunk('hello world')
    expect(h1).toBe(h2)
  })

  it('produces different hashes for different content', () => {
    const h1 = hashChunk('hello world')
    const h2 = hashChunk('different content')
    expect(h1).not.toBe(h2)
  })

  it('returns a 64-character hex string', () => {
    const h = hashChunk('test')
    expect(h).toMatch(/^[a-f0-9]{64}$/)
  })
})

describe('computeChunkDeltas', () => {
  it('marks all chunks as new when no manifest exists', () => {
    const chunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
      makeChunk('gamma content', 2),
    ]

    const deltas = computeChunkDeltas(chunks, undefined)

    expect(deltas).toHaveLength(3)
    for (const d of deltas) {
      expect(d.category).toBe('new')
      expect(d.hash).toMatch(/^[a-f0-9]{64}$/)
    }
  })

  it('marks all chunks as unchanged when hashes match', () => {
    const chunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
    ]

    const manifest = buildChunkManifest('doc1hash', chunks)

    const deltas = computeChunkDeltas(chunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const added = deltas.filter((d) => d.category === 'new')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(2)
    expect(added).toHaveLength(0)
    expect(removed).toHaveLength(0)
  })

  it('detects added chunks by hash not in old manifest', () => {
    const oldChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
    ]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const newChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
      makeChunk('new paragraph', 2),
    ]

    const deltas = computeChunkDeltas(newChunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const added = deltas.filter((d) => d.category === 'new')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(2)
    expect(added).toHaveLength(1)
    expect(added[0]?.chunk.content).toBe('new paragraph')
    expect(removed).toHaveLength(0)
  })

  it('detects removed chunks by hash in old but not new', () => {
    const oldChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
    ]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const newChunks: Chunk[] = [makeChunk('alpha content', 0)]

    const deltas = computeChunkDeltas(newChunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(1)
    expect(removed).toHaveLength(1)
    expect(removed[0]?.hash).toBe(hashChunk('beta content'))
  })

  it('marks reordered chunks as unchanged (same hashes, different positions)', () => {
    const oldChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
      makeChunk('gamma content', 2),
    ]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const newChunks: Chunk[] = [
      makeChunk('gamma content', 0),
      makeChunk('alpha content', 1),
      makeChunk('beta content', 2),
    ]

    const deltas = computeChunkDeltas(newChunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const added = deltas.filter((d) => d.category === 'new')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(3)
    expect(added).toHaveLength(0)
    expect(removed).toHaveLength(0)
  })

  it('handles complete document rewrite (all new hashes)', () => {
    const oldChunks: Chunk[] = [
      makeChunk('old alpha', 0),
      makeChunk('old beta', 1),
    ]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const newChunks: Chunk[] = [
      makeChunk('new alpha', 0),
      makeChunk('new beta', 1),
      makeChunk('new gamma', 2),
    ]

    const deltas = computeChunkDeltas(newChunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const added = deltas.filter((d) => d.category === 'new')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(0)
    expect(added).toHaveLength(3)
    expect(removed).toHaveLength(2)
  })

  it('handles one chunk changed while others remain identical', () => {
    const oldChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
      makeChunk('gamma content', 2),
    ]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const newChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta MODIFIED content', 1),
      makeChunk('gamma content', 2),
    ]

    const deltas = computeChunkDeltas(newChunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const added = deltas.filter((d) => d.category === 'new')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(2)
    expect(added).toHaveLength(1)
    expect(added[0]?.chunk.content).toBe('beta MODIFIED content')
    expect(removed).toHaveLength(1)
    expect(removed[0]?.hash).toBe(hashChunk('beta content'))
  })

  it('handles duplicate content chunks correctly', () => {
    const oldChunks: Chunk[] = [makeChunk('repeated', 0)]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const newChunks: Chunk[] = [makeChunk('repeated', 0), makeChunk('repeated', 1)]

    const deltas = computeChunkDeltas(newChunks, manifest)

    const unchanged = deltas.filter((d) => d.category === 'unchanged')
    const added = deltas.filter((d) => d.category === 'new')
    const removed = deltas.filter((d) => d.category === 'removed')

    expect(unchanged).toHaveLength(1)
    expect(added).toHaveLength(1)
    expect(removed).toHaveLength(0)
  })

  it('returns empty when both inputs are empty', () => {
    const deltas = computeChunkDeltas([], undefined)
    expect(deltas).toHaveLength(0)
  })

  it('marks all old chunks as removed when new set is empty', () => {
    const oldChunks: Chunk[] = [
      makeChunk('alpha content', 0),
      makeChunk('beta content', 1),
    ]
    const manifest = buildChunkManifest('doc1hash', oldChunks)

    const deltas = computeChunkDeltas([], manifest)

    const removed = deltas.filter((d) => d.category === 'removed')
    expect(removed).toHaveLength(2)
  })
})

describe('buildChunkManifest', () => {
  it('builds entries with correct hash and chunkId', () => {
    const chunks: Chunk[] = [makeChunk('chunk A', 0), makeChunk('chunk B', 1)]

    const manifest = buildChunkManifest('hash123', chunks)

    expect(manifest.documentHash).toBe('hash123')
    expect(manifest.chunks).toHaveLength(2)
    expect(manifest.chunks[0]?.hash).toBe(hashChunk('chunk A'))
    expect(manifest.chunks[0]?.chunkId).toBe('hash123_0')
    expect(manifest.chunks[1]?.hash).toBe(hashChunk('chunk B'))
    expect(manifest.chunks[1]?.chunkId).toBe('hash123_1')
  })
})

describe('readChunkManifest/writeChunkManifest', () => {
  const stores: Store[] = []

  afterEach(async () => {
    while (stores.length > 0) {
      const s = stores.pop()
      if (s !== undefined) await s.close()
    }
  })

  const freshStore = (): Store => {
    const s = createMemStore()
    stores.push(s)
    return s
  }

  it('round-trips through store correctly', async () => {
    const store = freshStore()
    const chunks: Chunk[] = [
      makeChunk('first chunk', 0),
      makeChunk('second chunk', 1),
      makeChunk('third chunk', 2),
    ]

    const manifest = buildChunkManifest('abc123def456', chunks)
    await writeChunkManifest(store, manifest)

    const loaded = await readChunkManifest(store, 'abc123def456')

    expect(loaded).not.toBeUndefined()
    expect(loaded?.documentHash).toBe('abc123def456')
    expect(loaded?.generation).toBe(1)
    expect(loaded?.chunks).toHaveLength(3)
    expect(loaded?.updatedAt).toMatch(/^\d{4}-\d{2}-\d{2}T/)

    for (let i = 0; i < chunks.length; i++) {
      const chunk = chunks[i]
      if (chunk === undefined) continue
      expect(loaded?.chunks[i]?.hash).toBe(hashChunk(chunk.content))
      expect(loaded?.chunks[i]?.chunkId).toBe(`abc123def456_${String(i)}`)
    }
  })

  it('returns undefined for missing manifest', async () => {
    const store = freshStore()
    const loaded = await readChunkManifest(store, 'nonexistent')
    expect(loaded).toBeUndefined()
  })

  it('increments generation on subsequent writes', async () => {
    const store = freshStore()
    const chunks1: Chunk[] = [makeChunk('content v1', 0)]
    const manifest1 = buildChunkManifest('docXYZ', chunks1)

    await writeChunkManifest(store, manifest1)
    const first = await readChunkManifest(store, 'docXYZ')
    expect(first?.generation).toBe(1)

    const chunks2: Chunk[] = [makeChunk('content v2', 0)]
    const manifest2 = buildChunkManifest('docXYZ', chunks2)

    await writeChunkManifest(store, manifest2)
    const second = await readChunkManifest(store, 'docXYZ')
    expect(second?.generation).toBe(2)
  })
})

File: go/knowledge/delta_test.go
// SPDX-License-Identifier: Apache-2.0
package knowledge

import (
	"context"
	"fmt"
	"testing"

	"github.com/jeffs-brain/memory/go/store/mem"
)

func TestComputeChunkDeltas_first_ingest(t *testing.T) {
	t.Parallel()
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "alpha content"},
		{DocumentID: "doc1", Ordinal: 1, Text: "beta content"},
		{DocumentID: "doc1", Ordinal: 2, Text: "gamma content"},
	}

	deltas := ComputeChunkDeltas(chunks, nil)

	if len(deltas) != 3 {
		t.Fatalf("expected 3 deltas, got %d", len(deltas))
	}
	for i, d := range deltas {
		if d.Category != DeltaNew {
			t.Errorf("delta[%d] expected DeltaNew, got %d", i, d.Category)
		}
		if d.Hash == "" {
			t.Errorf("delta[%d] has empty hash", i)
		}
	}
}

func TestComputeChunkDeltas_no_change(t *testing.T) {
	t.Parallel()
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "alpha content"},
		{DocumentID: "doc1", Ordinal: 1, Text: "beta content"},
	}

	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("alpha content"), ChunkID: "doc1hash_0"},
			{Hash: HashChunk("beta content"), ChunkID: "doc1hash_1"},
		},
	}

	deltas := ComputeChunkDeltas(chunks, manifest)

	var newCount, unchangedCount, removedCount int
	for _, d := range deltas {
		switch d.Category {
		case DeltaNew:
			newCount++
		case DeltaUnchanged:
			unchangedCount++
		case DeltaRemoved:
			removedCount++
		}
	}

	if newCount != 0 {
		t.Errorf("expected 0 new, got %d", newCount)
	}
	if unchangedCount != 2 {
		t.Errorf("expected 2 unchanged, got %d", unchangedCount)
	}
	if removedCount != 0 {
		t.Errorf("expected 0 removed, got %d", removedCount)
	}
}

func TestComputeChunkDeltas_chunk_added(t *testing.T) {
	t.Parallel()
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "alpha content"},
		{DocumentID: "doc1", Ordinal: 1, Text: "beta content"},
		{DocumentID: "doc1", Ordinal: 2, Text: "new paragraph"},
	}

	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("alpha content"), ChunkID: "doc1hash_0"},
			{Hash: HashChunk("beta content"), ChunkID: "doc1hash_1"},
		},
	}

	deltas := ComputeChunkDeltas(chunks, manifest)

	var newCount, unchangedCount int
	for _, d := range deltas {
		switch d.Category {
		case DeltaNew:
			newCount++
			if d.Chunk.Text != "new paragraph" {
				t.Errorf("expected new chunk text 'new paragraph', got %q", d.Chunk.Text)
			}
		case DeltaUnchanged:
			unchangedCount++
		case DeltaRemoved:
			t.Errorf("unexpected removed chunk: %+v", d)
		}
	}

	if newCount != 1 {
		t.Errorf("expected 1 new, got %d", newCount)
	}
	if unchangedCount != 2 {
		t.Errorf("expected 2 unchanged, got %d", unchangedCount)
	}
}

func TestComputeChunkDeltas_chunk_removed(t *testing.T) {
	t.Parallel()
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "alpha content"},
	}

	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("alpha content"), ChunkID: "doc1hash_0"},
			{Hash: HashChunk("beta content"), ChunkID: "doc1hash_1"},
		},
	}

	deltas := ComputeChunkDeltas(chunks, manifest)

	var unchangedCount, removedCount int
	for _, d := range deltas {
		switch d.Category {
		case DeltaNew:
			t.Errorf("unexpected new chunk: %+v", d)
		case DeltaUnchanged:
			unchangedCount++
		case DeltaRemoved:
			removedCount++
			if d.Hash != HashChunk("beta content") {
				t.Errorf("expected removed hash for 'beta content', got %q", d.Hash)
			}
		}
	}

	if unchangedCount != 1 {
		t.Errorf("expected 1 unchanged, got %d", unchangedCount)
	}
	if removedCount != 1 {
		t.Errorf("expected 1 removed, got %d", removedCount)
	}
}

func TestComputeChunkDeltas_reordered(t *testing.T) {
	t.Parallel()
	// Same content in different positions should all be DeltaUnchanged.
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "beta content"},
		{DocumentID: "doc1", Ordinal: 1, Text: "alpha content"},
		{DocumentID: "doc1", Ordinal: 2, Text: "gamma content"},
	}

	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("alpha content"), ChunkID: "doc1hash_0"},
			{Hash: HashChunk("beta content"), ChunkID: "doc1hash_1"},
			{Hash: HashChunk("gamma content"), ChunkID: "doc1hash_2"},
		},
	}

	deltas := ComputeChunkDeltas(chunks, manifest)

	for i, d := range deltas {
		if d.Category != DeltaUnchanged {
			t.Errorf("delta[%d] expected DeltaUnchanged (reorder), got %d for text %q", i, d.Category, d.Chunk.Text)
		}
	}
}

func TestComputeChunkDeltas_one_chunk_changed(t *testing.T) {
	t.Parallel()
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "alpha content"},
		{DocumentID: "doc1", Ordinal: 1, Text: "beta MODIFIED content"},
		{DocumentID: "doc1", Ordinal: 2, Text: "gamma content"},
	}

	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("alpha content"), ChunkID: "doc1hash_0"},
			{Hash: HashChunk("beta content"), ChunkID: "doc1hash_1"},
			{Hash: HashChunk("gamma content"), ChunkID: "doc1hash_2"},
		},
	}

	deltas := ComputeChunkDeltas(chunks, manifest)

	var newCount, unchangedCount, removedCount int
	for _, d := range deltas {
		switch d.Category {
		case DeltaNew:
			newCount++
			if d.Chunk.Text != "beta MODIFIED content" {
				t.Errorf("expected new chunk 'beta MODIFIED content', got %q", d.Chunk.Text)
			}
		case DeltaUnchanged:
			unchangedCount++
		case DeltaRemoved:
			removedCount++
			if d.Hash != HashChunk("beta content") {
				t.Errorf("expected removed hash for 'beta content', got %q", d.Hash)
			}
		}
	}

	if newCount != 1 {
		t.Errorf("expected 1 new, got %d", newCount)
	}
	if unchangedCount != 2 {
		t.Errorf("expected 2 unchanged, got %d", unchangedCount)
	}
	if removedCount != 1 {
		t.Errorf("expected 1 removed, got %d", removedCount)
	}
}

func TestComputeChunkDeltas_empty_new_chunks(t *testing.T) {
	t.Parallel()
	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("alpha content"), ChunkID: "doc1hash_0"},
			{Hash: HashChunk("beta content"), ChunkID: "doc1hash_1"},
		},
	}

	deltas := ComputeChunkDeltas(nil, manifest)

	var removedCount int
	for _, d := range deltas {
		if d.Category != DeltaRemoved {
			t.Errorf("expected DeltaRemoved, got %d", d.Category)
		}
		removedCount++
	}
	if removedCount != 2 {
		t.Errorf("expected 2 removed, got %d", removedCount)
	}
}

func TestComputeChunkDeltas_duplicate_content(t *testing.T) {
	t.Parallel()
	// Two chunks with identical content in new set, one in old manifest.
	// One should be unchanged, one should be new.
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "repeated"},
		{DocumentID: "doc1", Ordinal: 1, Text: "repeated"},
	}

	manifest := &ChunkManifest{
		DocumentHash: "doc1hash",
		Generation:   1,
		Chunks: []ChunkManifestEntry{
			{Hash: HashChunk("repeated"), ChunkID: "doc1hash_0"},
		},
	}

	deltas := ComputeChunkDeltas(chunks, manifest)

	var newCount, unchangedCount int
	for _, d := range deltas {
		switch d.Category {
		case DeltaNew:
			newCount++
		case DeltaUnchanged:
			unchangedCount++
		case DeltaRemoved:
			t.Errorf("unexpected removed delta")
		}
	}

	if unchangedCount != 1 {
		t.Errorf("expected 1 unchanged, got %d", unchangedCount)
	}
	if newCount != 1 {
		t.Errorf("expected 1 new, got %d", newCount)
	}
}

func TestReadWriteChunkManifest_round_trip(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()
	t.Cleanup(func() { _ = store.Close() })

	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "first chunk"},
		{DocumentID: "doc1", Ordinal: 1, Text: "second chunk"},
		{DocumentID: "doc1", Ordinal: 2, Text: "third chunk"},
	}

	manifest := BuildChunkManifest("abc123def456", chunks)

	if err := WriteChunkManifest(ctx, store, manifest); err != nil {
		t.Fatalf("write manifest: %v", err)
	}

	loaded, err := ReadChunkManifest(ctx, store, "abc123def456")
	if err != nil {
		t.Fatalf("read manifest: %v", err)
	}
	if loaded == nil {
		t.Fatal("expected non-nil manifest after write")
	}

	if loaded.DocumentHash != "abc123def456" {
		t.Errorf("expected document hash 'abc123def456', got %q", loaded.DocumentHash)
	}
	if loaded.Generation != 1 {
		t.Errorf("expected generation 1, got %d", loaded.Generation)
	}
	if len(loaded.Chunks) != 3 {
		t.Fatalf("expected 3 chunks, got %d", len(loaded.Chunks))
	}

	for i, entry := range loaded.Chunks {
		expectedHash := HashChunk(chunks[i].Text)
		if entry.Hash != expectedHash {
			t.Errorf("chunk[%d] hash mismatch: got %q, want %q", i, entry.Hash, expectedHash)
		}
		expectedID := fmt.Sprintf("abc123def456_%d", i)
		if entry.ChunkID != expectedID {
			t.Errorf("chunk[%d] id mismatch: got %q, want %q", i, entry.ChunkID, expectedID)
		}
	}

	if loaded.UpdatedAt.IsZero() {
		t.Error("expected non-zero UpdatedAt")
	}
}

func TestReadChunkManifest_missing(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()
	t.Cleanup(func() { _ = store.Close() })

	loaded, err := ReadChunkManifest(ctx, store, "nonexistent")
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if loaded != nil {
		t.Fatalf("expected nil manifest for missing document, got %+v", loaded)
	}
}

func TestWriteChunkManifest_increments_generation(t *testing.T) {
	t.Parallel()
	ctx := context.Background()
	store := mem.New()
	t.Cleanup(func() { _ = store.Close() })

	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "content v1"},
	}

	manifest := BuildChunkManifest("docXYZ", chunks)

	if err := WriteChunkManifest(ctx, store, manifest); err != nil {
		t.Fatalf("first write: %v", err)
	}

	loaded, err := ReadChunkManifest(ctx, store, "docXYZ")
	if err != nil {
		t.Fatalf("first read: %v", err)
	}
	if loaded.Generation != 1 {
		t.Errorf("expected generation 1 after first write, got %d", loaded.Generation)
	}

	// Write again with different content.
	chunks2 := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "content v2"},
	}
	manifest2 := BuildChunkManifest("docXYZ", chunks2)

	if err := WriteChunkManifest(ctx, store, manifest2); err != nil {
		t.Fatalf("second write: %v", err)
	}

	loaded2, err := ReadChunkManifest(ctx, store, "docXYZ")
	if err != nil {
		t.Fatalf("second read: %v", err)
	}
	if loaded2.Generation != 2 {
		t.Errorf("expected generation 2 after second write, got %d", loaded2.Generation)
	}
}

func TestHashChunk_deterministic(t *testing.T) {
	t.Parallel()
	h1 := HashChunk("hello world")
	h2 := HashChunk("hello world")
	if h1 != h2 {
		t.Errorf("HashChunk not deterministic: %q != %q", h1, h2)
	}
	h3 := HashChunk("different content")
	if h1 == h3 {
		t.Error("HashChunk collision on different content")
	}
}

func TestBuildChunkManifest(t *testing.T) {
	t.Parallel()
	chunks := []Chunk{
		{DocumentID: "doc1", Ordinal: 0, Text: "chunk A"},
		{DocumentID: "doc1", Ordinal: 1, Text: "chunk B"},
	}

	manifest := BuildChunkManifest("hash123", chunks)

	if manifest.DocumentHash != "hash123" {
		t.Errorf("expected document hash 'hash123', got %q", manifest.DocumentHash)
	}
	if len(manifest.Chunks) != 2 {
		t.Fatalf("expected 2 entries, got %d", len(manifest.Chunks))
	}
	if manifest.Chunks[0].ChunkID != "hash123_0" {
		t.Errorf("expected chunk id 'hash123_0', got %q", manifest.Chunks[0].ChunkID)
	}
	if manifest.Chunks[1].ChunkID != "hash123_1" {
		t.Errorf("expected chunk id 'hash123_1', got %q", manifest.Chunks[1].ChunkID)
	}
	if manifest.Chunks[0].Hash != HashChunk("chunk A") {
		t.Errorf("chunk[0] hash mismatch")
	}
	if manifest.Chunks[1].Hash != HashChunk("chunk B") {
		t.Errorf("chunk[1] hash mismatch")
	}
}
